### ISSUE TITLE
Wrong error when typechecking complex function argument return type

### ISSUE THREAD

**Issue by aralisza** [Microsoft/TypeScript#23531]
This issue was discovered in React setState.

**TypeScript Version:**  2.8.1

**Search Terms:** typescript wrong error react setstate type

**Code**

```ts
interface UploadState {
    acceptedFile?: File;
    uploadedId?: string;
    progress: number | null;
}

interface SomeClass<S> {
    setState<K extends keyof S>(
        state: (() => (Pick<S, K> | S | null)) | (Pick<S, K> | S | null),
    ): void;
}

class Test implements SomeClass<UploadState>{
    setState<K extends keyof UploadState>(state: (() => (Pick<UploadState, K> | UploadState | null)) | (Pick<UploadState, K> | UploadState | null)) {

    }
}

new Test().setState(() => {
    return {
        uploadedId: "test",
        progress: 3,
        acceptedFile: 3
    };
});
```
**Expected behavior:**
Typescript should tell me that acceptedFile should either be a File or undefined. 3 is not an acceptable value.
**Actual behavior:**
Typescript says:
```
Argument of type '() => { uploadedId: string; progress: number; acceptedFile: number; }' is not assignable to parameter of type 'UploadState | (() => UploadState | Pick<UploadState, "acceptedFile" | "uploadedId" | "progress"> ...'.
  Type '() => { uploadedId: string; progress: number; acceptedFile: number; }' is not assignable to type 'Pick<UploadState, "acceptedFile" | "uploadedId" | "progress">'.
    Property 'progress' is missing in type '() => { uploadedId: string; progress: number; acceptedFile: number; }'.
function(): {
    uploadedId: string;
    progress: number;
    acceptedFile: number;
}
```
**Playground Link:**  
[playground link](https://www.typescriptlang.org/play/#src=interface%20UploadState%20%7B%0D%0A%20%20%20%20acceptedFile%3F%3A%20File%3B%0D%0A%20%20%20%20uploadedId%3F%3A%20string%3B%0D%0A%20%20%20%20progress%3A%20number%20%7C%20null%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20SomeClass%3CS%3E%20%7B%0D%0A%20%20%20%20setState%3CK%20extends%20keyof%20S%3E(%0D%0A%20%20%20%20%20%20%20%20state%3A%20(()%20%3D%3E%20(Pick%3CS%2C%20K%3E%20%7C%20S%20%7C%20null))%20%7C%20(Pick%3CS%2C%20K%3E%20%7C%20S%20%7C%20null)%2C%0D%0A%20%20%20%20)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aclass%20Test%20implements%20SomeClass%3CUploadState%3E%7B%0D%0A%20%20%20%20setState%3CK%20extends%20keyof%20UploadState%3E(state%3A%20(()%20%3D%3E%20(Pick%3CUploadState%2C%20K%3E%20%7C%20UploadState%20%7C%20null))%20%7C%20(Pick%3CUploadState%2C%20K%3E%20%7C%20UploadState%20%7C%20null))%20%7B%0D%0A%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Anew%20Test().setState(()%20%3D%3E%20%7B%0D%0A%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20uploadedId%3A%20%22test%22%2C%0D%0A%20%20%20%20%20%20%20%20progress%3A%203%2C%0D%0A%20%20%20%20%20%20%20%20acceptedFile%3A%203%0D%0A%20%20%20%20%7D%3B%0D%0A%7D)%3B)

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
Given a union type `X | Y | Z` and an unassignable source type `T`, we arbitrarily pick one of `X, `Y`, or `Z` to demonstrate that `T` isn't assignable to it. In this case the compiler is complaining that your function is missing a property, rather than the function's return type being missing the property. This is not a *wrong* error, it's just not the *best* error.

The challenge of picking the "best" target type to give the "best" error is a shaky business of heuristics - we can't show you what *every* error would look like (there may be dozens of members in the union)

@sandersn had done some work in the past to fish out better targets, but I'm not sure if that logic is applicable to this situation?

**Comment by sandersn** [comment]
No, it only applies to discriminated unions. The compiler has no code to choose the "best match" between `T | () => T` when reporting errors.

If we could come up with a understandable and relatively simple heuristic for best match, it might be worth implementing since we don't care as much about performance in the error case. Suggestions and discussion are welcome. Keep in mind the following constraints:

In case the best match isn't clear, or isn't possible to calculate, the proposed algorithm should be:
1. stable. Small changes to the union shouldn't flip between several different error messages.
1. obvious. An expert should be able to figure out why that member of the union was chosen, without needing to know the internals of the implementation.
1. simple. An implementer who reads the code should be confident that the algorithm won't crash, take forever, or return wrong errors (opposed to not-best errors like we have today).

**Comment by RyanCavanaugh** [comment]
> My main complaint with this error message is that it is telling me I am missing property progress in the function's return even though it wasn't missing.

This isn't what the error says. It says 
```
Argument of type '() => { uploadedId: string; progress: number; acceptedFile: number; }' is not assignable to parameter of type 'UploadState | (() => UploadState | Pick<UploadState, "acceptedFile" | "uploadedId" | "progress"> ...'.
  Type '() => { uploadedId: string; progress: number; acceptedFile: number; }' is not assignable to type 'Pick<UploadState, "acceptedFile" | "uploadedId" | "progress">'.
    Property 'progress' is missing in type '() => { uploadedId: string; progress: number; acceptedFile: number; }'.
function(): {
```
It's saying that `progress` is missing *on the function itself*, not that it's missing in the return value. Which is technically correct, just not very helpful.
